lib/eclipse/EclipseState/Schedule/MSW/Segment.cpp

```diff
@@ -146,4 +150,16 @@ namespace Opm {
             && this->m_volume            == rhs.m_volume
             && this->m_data_ready        == rhs.m_data_ready;
     }
+
+    void Segment::updateSpiralICD(const SpiralICD& spiral_icd) {
+        m_segment_type = WellSegment::SPIRALICD;
+        m_spiral_icd = std::make_shared<SpiralICD>(spiral_icd);
+    }
+
+    const std::shared_ptr<const SpiralICD>& Segment::spiralICD() const {
```

Comment:
opm-parser has been on a quite strict `shared_ptr<>` diet; do you really need a `std::shared_ptr` here - or are you just following the old opm-parser pattern of excessive `shared_ptr` abuse?

Reply:
Well - first of all; from the outset opm-parser used `std::shared_ptr<T>` extensively; mostly in situations where *shared ownership* - was not relevant at all. Mainly due to @jokva the code base was put on a strict `std::shared_ptr<T>` diet - mostly replacing with references and automatically initialized stack variables - that was a great improvement.

My personal take on this is:

1. Use automatically initialized stack variables.
2. Try harder to use automatically initialized stack variables.
3. Use `std::unique_ptr<T>` for situations where you might - or might not - have an object.
4. Use `std::shared_ptr<T>` for situations where you actually have shared ownership.

I have not looked in detail on your code, and maybe it is actually appropriate to use the `std::shared_ptr<>` - I will look; feel free to explain briefly.

The problem with using these smart pointers in the API is that they spread to consumer code; i.e. the library implementer affects the memory model used by the application developer - that is kind of ugly.

Reply:
Yes. The natural choice is 3. `std::unique_ptr`, since it is not something  related to all segments, only exists for certain segments.  But `std::fill` used in `DynamicState` `update` make it better to use `std::shared_ptr`. 

And also, in the schedule related, there are quite some of places, we get a Segment, or SegmentSet, modify some, then replace the old one. Although we can use `std::move` to avoid the problem related to `std::unique_ptr`, it turned out to be not very easy. 

Anyway, I might have some wrong understanding here, please provides some suggestions.

Reply:
Then you stick to `std::shared_ptr<>` - no problem!